### PR TITLE
Merge Communicator and CommunicatorI in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR merges these two classes. 

It's a purely cleanup PR, with some small reshuffling. The logic remains the same.

### CODE REVIEW

File: cpp/include/Ice/Initialize.h [hunk 1]
```diff
@@ -661,6 +662,32 @@ class ICE_API CommunicatorHolder
     CommunicatorPtr _communicator;
 };
 
+/**
+ * The output mode for xxxToString method such as identityToString and proxyToString. The actual encoding format for
+ * the string is the same for all modes: you don't need to specify an encoding format or mode when reading such a
+ * string.
+ */
+enum class ToStringMode : unsigned char
```
Comment (bernardnormier) [inline cpp/include/Ice/Initialize.h:670]: Moved here to "fix" a circular dependency between Communicator.h and Initialize.h.

File: cpp/src/Ice/Communicator.cpp [hunk 2]
```diff
@@ -30,3 +36,297 @@ Ice::Communicator::flushBatchRequestsAsync(CompressBatch compress)
         });
     return promise->get_future();
 }
+
+void
+Ice::Communicator::destroy() noexcept
```
Comment (bernardnormier) [inline cpp/src/Ice/Communicator.cpp:41]: All this code comes from CommunicatorI.cpp. The only difference is there is no "I".

File: cpp/src/Ice/CommunicatorFlushBatchAsync.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+#include "ConnectionI.h"
+#include "ConnectionFactory.h"
+#include "Instance.h"
+#include "ObjectAdapterFactory.h"
+
+using namespace std;
+using namespace Ice;
+using namespace IceInternal;
+
+CommunicatorFlushBatchAsync::~CommunicatorFlushBatchAsync()
```
Comment (bernardnormier) [inline cpp/src/Ice/CommunicatorFlushBatchAsync.cpp:16]: This code used to be in CommunicatorI.cpp. It didn't change.

File: cpp/src/Ice/CommunicatorFlushBatchAsync.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#ifndef ICE_COMMUNICATOR_FLUSH_BATCH_ASYNC_H
+#define ICE_COMMUNICATOR_FLUSH_BATCH_ASYNC_H
+
+#include <Ice/Communicator.h>
+#include "Ice/OutgoingAsync.h"
+
+namespace IceInternal
```
Comment (bernardnormier) [inline cpp/src/Ice/CommunicatorFlushBatchAsync.h:11]: Moved from CommunicatorI.h without change.

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]
Looks good!

**Pull request merged**